svm/src/account_loader.rs

```diff
@@ -131,45 +150,156 @@ impl<'a, CB: TransactionProcessingCallback> AccountLoader<'a, CB> {
             .feature_set
             .is_active(&feature_set::disable_account_loader_special_case::id());
 
-        if let Some(account_override) = self
-            .account_overrides
-            .and_then(|overrides| overrides.get(account_key))
-        {
-            Some(LoadedTransactionAccount {
-                loaded_size: account_override.data().len(),
-                account: account_override.clone(),
-                rent_collected: 0,
-            })
-        } else if let Some(program) = (use_program_cache && is_invisible_read)
+        if let Some(program) = (use_program_cache && is_invisible_read)
             .then_some(())
             .and_then(|_| self.program_cache.find(account_key))
         {
             // Optimization to skip loading of accounts which are only used as
             // programs in top-level instructions and not passed as instruction accounts.
-            Some(LoadedTransactionAccount {
+            return Some(LoadedTransactionAccount {
                 loaded_size: program.account_size,
                 account: account_shared_data_from_program(account_key, &self.program_accounts)
                     .ok()?,
                 rent_collected: 0,
-            })
+            });
+        }
+
+        let account = if let Some(cache_item) = self.account_cache.get_mut(account_key) {
+            // Inspecting a read-only account is a no-op. But inspecting a writable
+            // account must be done before rent collection, and must never be done
+            // again in-batch, because inspection is intended to preserve the
+            // state prior to any writes.
+            if is_writable && !cache_item.inspected_as_writable {
+                self.callbacks.inspect_account(
+                    account_key,
+                    AccountState::Alive(&cache_item.account),
+                    is_writable,
+                );
+                cache_item.inspected_as_writable = cache_item.inspected_as_writable || is_writable;
+            }
+
+            // If lamports is 0, a previous transaction deallocated this account.
+            // Return None without inspecting, so it can be recreated.
+            if cache_item.account.lamports() == 0 {
+                None
+            } else {
+                Some(cache_item.account.clone())
+            }
```

Comment:
while i was working on this i realized we dont need to store `AccountSharedData::default()` at all, so that saves us some `Arc::new()` calls. `account_cache` is private and `load_account()` is the only function that reads from it